Sources/StreamChat/Controllers/ThreadListController/ThreadListController.swift

```diff
@@ -0,0 +1,211 @@
+    internal private(set) var hasLoadedAllOlderThreads: Bool = false
+
+    /// A type-erased delegate.
+    var multicastDelegate: MulticastDelegate<ChatThreadListControllerDelegate> = .init() {
+        didSet {
+            stateMulticastDelegate.set(mainDelegate: multicastDelegate.mainDelegate)
+            stateMulticastDelegate.set(additionalDelegates: multicastDelegate.additionalDelegates)
+
+            // After setting delegate local changes will be fetched and observed.
+            startThreadListObserverIfNeeded()
+        }
+    }
+
+    private(set) lazy var threadListObserver: ListDatabaseObserverWrapper<ChatThread, ThreadDTO> = {
+        let request = ThreadDTO.threadListFetchRequest()
+        let observer = self.environment.createThreadListDatabaseObserver(
+            StreamRuntimeCheck._isBackgroundMappingEnabled,
+            client.databaseContainer,
+            request,
+            { try $0.asModel() }
+        )
+
+        observer.onDidChange = { [weak self] changes in
+            self?.delegateCallback { [weak self] in
+                guard let self = self else {
+                    return
+                }
+                $0.controller(self, didChangeThreads: changes)
+            }
+        }
+
+        return observer
+    }()
+
+    var _basePublishers: Any?
+    /// An internal backing object for all publicly available Combine publishers. We use it to simplify the way we expose
+    /// publishers. Instead of creating custom `Publisher` types, we use `CurrentValueSubject` and `PassthroughSubject` internally,
+    /// and expose the published values by mapping them to a read-only `AnyPublisher` type.
+    @available(iOS 13, *)
+    var basePublishers: BasePublishers {
+        if let value = _basePublishers as? BasePublishers {
+            return value
+        }
+        _basePublishers = BasePublishers(controller: self)
+        return _basePublishers as? BasePublishers ?? .init(controller: self)
+    }
+
+    private let environment: Environment
+
+    /// Creates a new `ChatThreadListController`.
+    init(
+        query: ThreadListQuery,
+        client: ChatClient,
+        environment: Environment = .init()
+    ) {
+        self.client = client
+        self.query = query
+        self.environment = environment
+    }
+
+    override internal func synchronize(_ completion: ((_ error: Error?) -> Void)? = nil) {
+        startThreadListObserverIfNeeded()
+
+        let limit = query.limit
+        updater.loadThreads(
+            query: query
+        ) { [weak self] result in
+            switch result {
+            case let .success(threadListResponse):
+                self?.threadListObserver.refreshItems { [weak self] in
+                    self?.callback {
+                        let threads = threadListResponse.threads
+                        self?.nextCursor = threadListResponse.next
+                        self?.state = .remoteDataFetched
```

Comment:
@laevandus I move this to inside the callback function, so for 99% of cases it will be thread safe, since these functions are called from the MainThread, and the callback will be also on the main thread (unless customers change this, which usually is not the case)